### PR TITLE
Implement a new swap price algorithm

### PR DESCRIPTION
**Pull request by amityadav0**
Swaps
The algorithm for processing assets swaps is given by:
`y = (x * Y * X) / (x + X)^2`, where x = input, X = Input Asset, Y = Output Asset, y = output

### CODE REVIEW

File: nearswap/src/internal.rs [hunk 1]
```diff
@@ -28,22 +28,12 @@ impl NearSwap {
...
     #[inline]
     pub(crate) fn calc_out_amount(&self, in_amount: u128, in_bal: u128, out_bal: u128) -> u128 {
-        // in_a * out_bal / (in_bal + in_a)  and scaling for fee
-        let in_net = u256::from(in_amount) * 997;
-        let r: u256 = in_net * u256::from(out_bal) / (u256::from(in_bal) * 1000 + in_net);
-        return r.as_u128();
-    }
+        // formula r = (in_amount * in_bal * out_bal ) / ( in_amount + in_bal ) ^ 2;
+        let denominator = (U512::from(in_amount) + U512::from(in_bal)) * U512::from(in_amount) + U512::from(in_bal);
+        let numerator = ( U512::from(in_amount) * U512::from(in_bal) * U512::from(out_bal));
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:33]: We don't need to use `U512`. Multiplying 2 `U128` numbers will never make a number bigger than `U256`.

`x * Y * X` could make a 512 bit number - but this won't happen, as the pool would need to consume more than a half of the liquidity of Y and X, and x will need to be an order of X.

File: nearswap/src/internal.rs [hunk 2]
```diff
@@ -76,51 +59,57 @@ impl NearSwap {
...
+    // calculateFee the fee of the swap
+    pub(crate) fn calcLiquidityFee(&self, X: u128, x: u128, Y: u128) -> u128 {
+        // ( x^2 *  Y ) / ( x + X )^2
+        let numerator = U512::from(x) * U512::from(x) * U512::from(Y);
+        let denominator = (U512::from(x) + U512::from(X)) * (U512::from(x) + U512::from(X));
+        if(denominator == U512::from(0)) {
+            return 0;
+        }
+        let r = numerator/denominator;
+        return r.as_u128();
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:74]:
```suggestion
    /// returns swap out amount and fee.
    pub(crate) fn calc_out_with_fee(&self, X: u128, x: u128, Y: u128) -> (u128, u128) {
        if x == 0 {
          return (0, 0);
        }
        fee = x*1000/3; // 0.3% x
        x = x-fee;
        (calc_out_amount(X, x, Y), fee)
```

File: nearswap/src/internal.rs [hunk 3]
```diff
@@ -76,51 +59,57 @@ impl NearSwap {
...
         ynear_in: Balance,
         token: &AccountId,
         tokens_out: Balance,
     ) {
+        let in_bal = p.ynear;
+        let out_bal = p.tokens;
+        let in_amount = ynear_in;
+
+        let liquidityFee = self.calcLiquidityFee(in_bal, in_amount, out_bal);
+        let amount_out = self.calc_out_amount(in_amount, in_bal, out_bal);
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply nearswap/src/internal.rs:91]: above we use `in_` and `out_` prefix - so let's be consistent here as well.

File: nearswap/src/internal.rs [hunk 4]
```diff
@@ -129,48 +118,75 @@ impl NearSwap {
...
         let user = env::predecessor_account_id();
+
+        let in_bal = p.tokens;
+        let out_bal = p.ynear;
+        let in_amount = token_in;
+
+        let liquidityFee = self.calcLiquidityFee(in_bal, in_amount, out_bal);
+        let amount_out = self.calc_out_amount(in_amount, in_bal, out_bal);
+        
+        assert!(amount_out >= ynear_out, "ERR_MIN_AMOUNT");
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:129]: let's use same update as above.
Also, let's move and define `"ERR_MIN_AMOUNT"` in errors.rs

File: nearswap/src/internal.rs [hunk 5]
```diff
@@ -76,51 +59,57 @@ impl NearSwap {
...
     }
 
+    // Should be at least tokens_out or swap will fail
+    // (prevents front running and other slippage issues).
     pub(crate) fn _swap_n2t(
         &mut self,
         p: &mut Pool,
         ynear_in: Balance,
         token: &AccountId,
         tokens_out: Balance,
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:84]:
```suggestion
        min_tokens_out: Balance,
```

File: nearswap/src/internal.rs [hunk 6]
```diff
@@ -76,51 +59,57 @@ impl NearSwap {
...
+        let denominator = (U512::from(x) + U512::from(X)) * (U512::from(x) + U512::from(X));
+        if(denominator == U512::from(0)) {
+            return 0;
+        }
+        let r = numerator/denominator;
+        return r.as_u128();
     }
 
+    // Should be at least tokens_out or swap will fail
+    // (prevents front running and other slippage issues).
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:78]:
```suggestion
    /// Should be at least `min_tokens_out` or swap will fail
    /// (prevents front running and other slippage issues).
```

File: nearswap/src/internal.rs [hunk 7]
```diff
@@ -129,48 +118,75 @@ impl NearSwap {
...
+        let amount_out_first = self.calc_out_amount(in_amount_first, in_bal_first, out_bal_first);
+
+        let in_bal_second = p2.ynear;
+        let out_bal_second = p2.tokens;
+        let in_amount_second = amount_out_first;
+
+        let liquidityFee_second = self.calcLiquidityFee(in_bal_second, in_amount_second, out_bal_second);
+        let amount_out_second = self.calc_out_amount(in_amount_second, in_bal_second, out_bal_second);
+
+        assert!(amount_out_second >= token2_out, "ERR_MIN_AMOUNT");
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:173]:
```suggestion
        let (swap_amount, _) = self.calc_out_with_fee(token1_in, p1.tokens, p1.ynear);
        let (out, _) = self.calc_out_with_fee(swap_amount, p2.ynear, p2.tokens);
        assert!(out >= min_token2_out, "ERR_MIN_AMOUNT");
```

Comment (robert-zaremba) [reply nearswap/src/internal.rs:173]: Let's also move `"ERR_MIN_AMOUNT"` to errors.

File: nearswap/src/internal.rs [hunk 8]
```diff
@@ -129,48 +118,75 @@ impl NearSwap {
...
     pub(crate) fn _swap_tokens(
         &mut self,
-        mut p1: Pool,
-        mut p2: Pool,
+        p1: &mut Pool,
+        p2: &mut Pool,
         token1: &AccountId,
         token1_in: Balance,
         token2: &AccountId,
         token2_out: Balance,
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:155]:
```suggestion
        min_token2_out: Balance,
```

File: nearswap/src/internal.rs [hunk 9]
```diff
@@ -129,48 +118,75 @@ impl NearSwap {
...
-        d.add(token2, token2_out);
+        d.remove(token1, in_amount_first);
+        d.add(token2, amount_out_second);
 
-        self.set_pool(&token1, &p1);
-        self.set_pool(&token2, &p2);
+        self.set_pool(&token1, p1);
+        self.set_pool(&token2, p2);
         self.set_deposit(&user, &d);
     }
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:191]: we need to return `out`

File: nearswap/src/util.rs [hunk 10]
```diff
@@ -25,6 +25,11 @@ construct_uint! {
     pub struct u256(4);
 }
 
+construct_uint! {
+    /// 512-bit unsigned integer.
+    pub struct U512(4);
+}
+
```
Comment (robert-zaremba) [inline nearswap/src/util.rs:32]: we can remove this.

File: nearswap/src/internal.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply nearswap/src/internal.rs:91]: let's remove lines 86-88 and use `p.ynear` ... directly

File: nearswap/Cargo.toml [hunk 11]
```diff
@@ -16,6 +16,5 @@ near-contract-standards = { git = "https://github.com/near/near-sdk-rs", rev = "
 uint = { version = "0.9.0", default-features = false }
 
 [dev-dependencies]
-near-primitives = { git = "https://github.com/nearprotocol/nearcore.git" }
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply nearswap/Cargo.toml:19]: ok

File: nearswap/src/internal.rs (hunk 1 above)
Comment (amityadav0) [reply nearswap/src/internal.rs:33]: ok

File: nearswap/src/internal.rs (hunk 4 above)
Comment (amityadav0) [reply nearswap/src/internal.rs:126]: ok

File: nearswap/src/internal.rs (hunk 9 above)
Comment (amityadav0) [reply nearswap/src/internal.rs:175]: ok

File: nearswap/src/util.rs (hunk 10 above)
Comment (amityadav0) [reply nearswap/src/util.rs:32]: removed

File: nearswap/src/internal.rs [hunk 12]
```diff
@@ -28,22 +28,12 @@ impl NearSwap {
...
     #[inline]
     pub(crate) fn calc_out_amount(&self, in_amount: u128, in_bal: u128, out_bal: u128) -> u128 {
-        // in_a * out_bal / (in_bal + in_a)  and scaling for fee
-        let in_net = u256::from(in_amount) * 997;
-        let r: u256 = in_net * u256::from(out_bal) / (u256::from(in_bal) * 1000 + in_net);
-        return r.as_u128();
-    }
+        // formula: y = (x * Y * X) / (x + X)^2
+        let denominator = (u256::from(in_amount) + u256::from(in_bal)) * u256::from(in_amount) + u256::from(in_bal);
+        let numerator = ( u256::from(in_amount) * u256::from(in_bal) * u256::from(out_bal));
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:33]:
```suggestion
        let x = u256::from(in_amount);
        let X = u256::from(in_bal);
        let numerator = ( x * u256::from(in_bal) * X);
        let mut denominator = (x + X);
        denominator *= denominator;
```

File: nearswap/src/internal.rs [hunk 13]
```diff
@@ -76,102 +59,122 @@ impl NearSwap {
...
-        let p_out = self.get_pool(&t_out);
-        let near_swap = self.calc_in_amount(tokens_out, p_out.ynear, p_out.tokens);
-        let tokens1_to_pay = self.calc_in_amount(near_swap, p_in.tokens, p_in.ynear);
-        println!(
-            "Swapping_out {} {} -> {} ynear -> {} {}",
-            tokens1_to_pay, t_in, near_swap, tokens_out, t_out
-        );
-        return (p_in, p_out, near_swap, tokens1_to_pay);
+    /// returns swap out amount and fee.
+    pub(crate) fn calc_out_with_fee(&self, mut x: u128,  X: u128, Y: u128) -> (u128, u128) {
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:66]: let's move this function up, just below `calc_out_amount`

File: nearswap/src/internal.rs [hunk 14]
```diff
@@ -76,102 +59,122 @@ impl NearSwap {
...
-    ) {
+        min_tokens_out: Balance,
+    ) -> Balance {
+        let in_bal = p.ynear;
+        let out_bal = p.tokens;
+        let in_amount = ynear_in;
+
+        let (out_amount, fee) = self.calc_out_with_fee(in_amount, in_bal, out_bal);
+
+        assert!(out_amount >= min_tokens_out, ERR25_MIN_AMOUNT);
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:90]:
```suggestion
        let (out_amount, fee) = self.calc_out_with_fee(in_amount, in_bal, out_bal);
        assert!(out_amount >= min_tokens_out, ERR25_MIN_AMOUNT);
```

File: nearswap/src/internal.rs [hunk 15]
```diff
@@ -76,102 +59,122 @@ impl NearSwap {
...
         d.remove_near(ynear_in);
-        d.add(token, tokens_out);
+        d.add(token, out_amount);
 
         self.set_pool(token, p);
         self.set_deposit(&user, &d);
+        out_amount
     }
 
+    // Should be at least ynear_out or swap will fail
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:109]:
```suggestion
    // Should be at least `min_ynear_out` or swap will fail
```

File: nearswap/src/internal.rs [hunk 16]
```diff
@@ -76,102 +59,122 @@ impl NearSwap {
...
+    // (prevents front running and other slippage issues).
     pub(crate) fn _swap_t2n(
         &mut self,
         p: &mut Pool,
         token: &AccountId,
         token_in: Balance,
         ynear_out: Balance,
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:116]:
```suggestion
        min_ynear_out: Balance,
```

File: nearswap/src/internal.rs [hunk 17]
```diff
@@ -76,102 +59,122 @@ impl NearSwap {
...
+    ) -> Balance {
         let user = env::predecessor_account_id();
+
+        let in_bal = p.tokens;
+        let out_bal = p.ynear;
+        let in_amount = token_in;
+
+        let (out_amount, fee) = self.calc_out_with_fee(in_amount, in_bal, out_bal);
+
+        assert!(out_amount >= ynear_out, ERR25_MIN_AMOUNT);
```
Comment (robert-zaremba) [inline nearswap/src/internal.rs:126]:
```suggestion
        let (out_amount, fee) = self.calc_out_with_fee(in_amount, in_bal, out_bal);
        assert!(out_amount >= ynear_out, ERR25_MIN_AMOUNT);
```

**Review by robert-zaremba — APPROVED** [review]
pre-approving, left few comments.

File: nearswap/src/internal.rs (hunk 13 above)
Comment (amityadav0) [reply nearswap/src/internal.rs:66]: ok

**Pull request merged**